### ISSUE TITLE
Different attributions for IntegratedGradients/LayerIntegratedGradients for BERT embeddings.

### ISSUE THREAD

**Issue by phiwi** [pytorch/captum#1132]
## ❓ Questions and Help

When doing inference on a trained `BertForSequenceClassification` model (which has ab [BertModel](https://huggingface.co/docs/transformers/v4.27.2/en/model_doc/bert#transformers.BertModel) as its base), I get slightly different results for 

1. `IntegratedGradients `and inputting embeddings
2. `LayerIntegratedGradients `initialized for the `model.bert.embeddings` layer and inputting input ids

In the following "ig" stands for IntegratedGradients and "lig" for LayerIntegratedgradients.

Forward functions:

```
def ig_func(inputs):
    pred = model(inputs)
    return pred.logits

def ig_embed_func(inputs):
    pred = model(inputs_embeds=inputs)
    return pred.logits
```

**Attribution class:**

```
lig = LayerIntegratedGradients(ig_func, model.bert.embeddings)
ig = IntegratedGradients(ig_embed_func)
```

**Input representations:**

```
input_ids = ...
ref_input_ids = token_reference.generate_reference(input_ids.size(-1), device=device).unsqueeze(0)
input_embeds = model.get_input_embeddings()(input_ids)
ref_input_embeds = model.get_input_embeddings()(ref_input_ids)
```

**Calculation:**

```
lig_attributions, delta = lig.attribute(
    inputs=input_ids,
    baselines=ref_input_ids,
    additional_forward_args=None,
    return_convergence_delta=True,
    n_steps=50,
    target=0
)
ig_attributions, delta = ig.attribute(
    inputs=input_embeds,
    baselines=ref_input_embeds,
    additional_forward_args=None,
    return_convergence_delta=True,
    n_steps=50,
    target=0
)
```

**Summation:**

```
ig_attributions_sum = ig_attributions.sum(dim=-1).squeeze(0)
lig_attributions_sum = lig_attributions.sum(dim=-1).squeeze(0)
```

**Output**

```
>>> ig_attributions_sum[:10], lig_attributions_sum[:10]
(tensor([ 0.0755, -0.0366, -0.0526,  0.1775,  0.0118,  0.0393,  0.0243,  0.0654,
          0.0490,  0.0587], dtype=torch.float64, grad_fn=<SliceBackward0>),
 tensor([ 0.0757, -0.0333, -0.0524,  0.0858,  0.0139,  0.0132,  0.0586,  0.0767,
          0.0621,  0.0593], dtype=torch.float64))
```

The outputs are "similar" but not identical.

To the best of my knowledge, the differences occur with any input and any other of our trained models.

**Comment by craymichael** [comment]
Apologies for the delay.

These two setups are close, but they are not exactly the same path.

With:

```python
IntegratedGradients(ig_embed_func)
```

you are interpolating the `inputs_embeds` tensor you pass to the model. Hugging Face BERT will still add position embeddings, token-type embeddings, layer norm, and dropout inside the embedding module.

With:

```python
LayerIntegratedGradients(ig_func, model.bert.embeddings)
```

you are attributing the output of the whole `BertEmbeddings` module, which includes word embeddings plus position / token-type embeddings and the embedding layer's normalization / dropout behavior.

To make the comparison closer:

```python
model.eval()
lig = LayerIntegratedGradients(ig_func, model.bert.embeddings.word_embeddings)
```

or use `attribute_to_layer_input=True` if you specifically want attribution to the input of the selected layer rather than its output.

Small differences can also come from dropout if the model is not in eval mode. Larger token-level differences are expected if one path attributes word embeddings only and the other attributes the combined BERT embedding output.

This should answer the question, so I'll close this for now. Please feel free to reply with any follow-up questions.
